packages/expo-clipboard/android/src/main/java/expo/modules/clipboard/ClipboardFileProvider.kt

```diff
@@ -0,0 +1,360 @@
+package expo.modules.clipboard
+
+import android.content.ContentProvider
+import android.content.ContentValues
+import android.content.Context
+import android.content.pm.PackageManager
+import android.content.pm.ProviderInfo
+import android.database.Cursor
+import android.database.MatrixCursor
+import android.net.Uri
+import android.os.Build
+import android.os.Environment
+import android.os.ParcelFileDescriptor
+import android.provider.OpenableColumns
+import android.text.TextUtils
+import android.webkit.MimeTypeMap
+import androidx.annotation.RequiresApi
+import androidx.core.content.FileProvider
+import org.xmlpull.v1.XmlPullParser.END_DOCUMENT
+import org.xmlpull.v1.XmlPullParser.START_TAG
+import org.xmlpull.v1.XmlPullParserException
+import java.io.File
+import java.io.FileNotFoundException
+import java.io.IOException
+
+/**
+ * This is a modified version of [FileProvider]
+ * that facilitates exposing files associated with an app by creating
+ * a `content://` uri without using Androids URI permission mechanism.
+ * In contrast to [FileProvider], this provider _must_ be exported.
+ *
+ * The difference is that [FileProvider] forbids provider to be _exported_
+ * which means it cannot easily grant access to any app installed
+ * on the device. This becomes even more problematic with API 31, when
+ * [PackageManager.getInstalledApplications] doesn't return all
+ * installed apps, so we cannot iterate and use [Context.grantUriPermission]
+ * easily
+ *
+ * For usage details, see [FileProvider] documentation
+ */
+@RequiresApi(Build.VERSION_CODES.KITKAT)
+class ClipboardFileProvider : ContentProvider() {
```

Comment:
I made this implementation as close as possible to the original [`FileProvider`](https://android.googlesource.com/platform/frameworks/support/+/android-support-lib-19.1.0/v4/java/android/support/v4/content/FileProvider.java) implementation, but it can be made much slimmer if needed, at the expense of flexibility.

Reply:
So, my understanding is it's a direct copy, but some part is changed right? If so please comment what part has been adapted 👀 
Or is it a complete rewrite? 🤔 
I'm a bit worried about the complexity of this class and that nobody would really be able to maintain after it's merged 😅 but I see that there's no other path to take to handle the images more or less correctly right? 🤔

Reply:
It's mostly a direct copy, but I kotlinized it and cleaned up a bit (reduced from 700+ lines to 360), significant changes to logic are in a few places:
- `attachInfo()` - original `FileProvider` cannot be exported
- `create/update/delete` methods - IIRC some of them were defined, but here these operations shouldn't be possible
- got rid of some pre-KitKat (Android <= 4.4) compatibility workarounds
- fixed some weird Java null handling 🤷 

Something very similar has been done in this lib: https://github.com/cketti/PublicFileProvider

If you really feel this is still too complex, I can wipe out the whole `PathStrategy` logic and all that XML parsing, leaving the hardcoded URI for `<app cache dir>/.clipboard/*.{png,jpg}`

Reply:
Can't we just subclass `FileProvider` and override what's problematic? 🤔 
To achieve possibility of using `<provider exported=true ... />` construction?

Reply:
We cannot, I tried that already, but proper overriding `attachInfo()` is impossible. I would need to call superclass 2 levels above, when omitting direct superclass call.

Reply:
Ok, I get it now. Given all the things you've described I need to admit that's the path we need to choose right now 👀 Thank you for all the clarification 👍